firedrake/slate/slate.py

```diff
@@ -140,6 +141,42 @@ def inv(self):
     def T(self):
         return Transpose(self)
 
+    def block(self, test_idx, trial_idx=None):
+        """Extract a particular block of a rank 1 or 2 tensor.
+
+        :arg test_idx: Indices of the test function space(s)
+                       This should be an integer or iterable of
+                       integers.
+        :arg trial_idx: Indices of the trial function space(s)
+                        This should be an integer or iterable of
+                        integers. If `None`, then the tensor
+                        is required to be rank 1.
+        """
+
+        def to_tuple(i):
+            try:
+                i = tuple(i)
+            except TypeError:
+                i = (i,)
+            return i
+
+        # One should not use this on rank-0 expressions
+        if self.rank == 0:
+            raise ValueError("Scalars do not have blocks.")
+
+        elif self.rank == 1:
+            if trial_idx is not None:
+                raise ValueError("Trial indices found for a 1-tensor. "
+                                 "If you're attempting to extract a block "
+                                 "over multiple test spaces, provide a tuple.")
+            arg_indices = (to_tuple(test_idx),)
+
+        else:
+            assert self.rank == 2, "Rank %d not supported." % self.rank
+            arg_indices = (to_tuple(test_idx), to_tuple(trial_idx))
+
+        return Block(self, arg_indices)
```

Comment:
I'd prefer to be liberal in what one receives here.  I think you should be allowed to `block` a 0-tensor, by passing `tensor.block(()) => tensor`.  This way you don't have to write special-case code outside.

Similarly, you could (like `ExtractSubBlock`) just accept any number of indices, but only handle the ones you care about.

Reply:
But if you're going to do all this type-checking, best to do it in `Block`, since then you get it when either the user calls `Tensor.block`, or else when they call `Block` directly.

Reply:
Okay I'll update this!